### ISSUE TITLE
Add webRTC support (audio and video calls)

### ISSUE THREAD

**Issue by nioc** [nioc/xmpp-web#1]

**Comment by wiktor-k** [comment]
This is actually easier than it sounds like and Stanza even has a demo for Jingle calls: https://github.com/legastero/stanza/blob/master/examples/jingle-demo.html

(I did verify that it works)

**Comment by nioc** [comment]
Thanks @wiktor-k I'll look at it.

**Comment by wiktor-k** [comment]
@nioc I actually spent some time thinking and writing small PoC JS XMPP clients if you'd like to I can send you my "braindump" somewhere... if not... then it's okay too :)

**Comment by nioc** [comment]
@wiktor-k Ok, with pleasure. May be you can publish it as a new Github repo?

**Comment by zsinba** [comment]
how about this work?   ??

**Comment by ThatOneCalculator** [comment]
With the switch to xmpp.js, WebRTC integration should be made even simpler, yea?

**Comment by ThatOneCalculator** [comment]
:eyes: https://github.com/otalk/jingle.js

**Comment by tsacks** [comment]
Just going to chime in and volunteer to help with this one.

The lack of clients with good voice, video, and screensharing support was brought to my attention after recent issues with Discord, so I've been looking for a project to try to help get these features into.

**Comment by tsacks** [comment]
I'm looking into how to do mermaid diagrams, and I should absolutely be able to go through make some diagrams to show how the XEPs go together, and what happens for each. Will get to them as soon as I can.

**Comment by nioc** [comment]
Top.
Here is a working example for any github/markdown document:
````
```mermaid
sequenceDiagram
    participant dotcom
    participant iframe
    participant viewscreen
    dotcom->>iframe: loads html w/ iframe url
    iframe->>viewscreen: request template
    viewscreen->>iframe: html & javascript
    iframe->>dotcom: iframe ready
    dotcom->>iframe: set mermaid data on iframe
    iframe->>iframe: render mermaid
```
````

Which renders like this:
```mermaid
sequenceDiagram
    participant dotcom
    participant iframe
    participant viewscreen
    dotcom->>iframe: loads html w/ iframe url
    iframe->>viewscreen: request template
    viewscreen->>iframe: html & javascript
    iframe->>dotcom: iframe ready
    dotcom->>iframe: set mermaid data on iframe
    iframe->>iframe: render mermaid
```
There are good codium/vscode plugin which allow live preview.

**Comment by tsacks** [comment]
The documentation for most of the XEPs use the example names of Romeo and Juliet or the three witches from Macbeth, depending on number of participants. I'll go ahead and use those in the diagrams.

In terms of how Jingle works, everything builds off of [XEP-0166: Jingle](https://xmpp.org/extensions/xep-0166.html), and every other XEP adds something to this process.

```mermaid
sequenceDiagram
    participant Romeo
    participant Juliet
    Romeo-->>Juliet: session-initiate
    Juliet-->>Romeo: ack
    Juliet-->>Romeo: session-accept
    Romeo-->>Juliet: ack
    Romeo<<->>Juliet: MEDIA SESSION
    Juliet-->>Romeo: session-terminate
    Romeo-->>Juliet: ack
```

Important note: either party can send a `session-terminate` message, the other sends the `ack`.

The next standards are [XEP-0167: Jingle RTP Sessions](https://xmpp.org/extensions/xep-0167.html) (it defines the first Application for using Jingle: voice and video), and [XEP-0176: Jingle ICE-UDP Transport Method](https://xmpp.org/extensions/xep-0176.html) (the transport to get that data from one peer to the other, in the most compatible way). I'm focusing on these two specifically, because they're the recommendations from the [XEP-0479: XMPP Compliance Suites 2023](https://xmpp.org/extensions/xep-0479.html). Once we add XEP-0167 and XEP-0176, the flow looks like this:

```mermaid
sequenceDiagram
    participant Romeo
    participant Juliet
    Romeo-->>Juliet: session-initiate
    Juliet-->>Romeo: ack
    Juliet-->>Romeo: session-info (ringing)
    Romeo-->>Juliet: ack
    Juliet-->>Romeo: session-accept
    Romeo-->>Juliet: ack
    Romeo<<-->>Juliet: [Optional transport and application negotiating]
    Romeo<<->>Juliet: STUN connectivity checks
    Romeo<<->>Juliet: Audio/Video (RTP)
    Juliet-->>Romeo: session-terminate
    Romeo-->>Juliet: ack
```

This is probably a good starting point. These are the requirements to get the voice working, and is a good target for passing the compliance tests.

**Comment by nioc** [comment]
I understand this is for a 1-1 conversation, right?
Does this applies also to MUC? I'm a bit confused about how multiple users should share their screen at the same time. I'll give a look to others clients.

**Comment by tsacks** [comment]
Yep, this is all for 1-1 chat. Those are the simpler ones to get working, and they're all that's required for compliance testing. 

By default, Jingle sends these messages as IQ stanzas to individual devices. XEP-0353 Jingle Message Initiation allows you to send a regular Message stanza to a contact first. You still follow it with the regular process, but messaging first let's you only have to send one IQ to one device, and allows every client to show when the call started and stopped.

MUC has to use XEP-0272 Multiparty Jingle. Instead of using message or IQ stanzas, muji does the coordinating through presence stanzas in the group chat. Essentially, everyone tells everyone else what ip they have, and what they're sharing. And everyone sets up a p2p session with everyone else.

**Comment by tsacks** [comment]
Small status update:

I've got it to the point where xmpp-web knows how to start ringing when called, and can tell the caller that we're ringing. It can also send either an 'accept' or 'decline' message.

Still need to do the actual negotiation and connecting the call, but it's a start.

I've been slow at getting things working because of lack of time to work on it, but I am making progress! It's gonna need a lot of cleaning up, but it's looking very possible from here.

Edit: demo of what I have so far is available at https://github.com/tsacks/xmpp-web/tree/experiment-jingle

**Comment by nioc** [comment]
In order to avoid noise on the issue (and notifications), can you create the associate pull request? We'll continue the discussion there 😉

This should allow me to push some changes in your forked branch (checkbox "_Allow edits from maintainers_" on PR not unchecked).

**Comment by nioc** [comment]
Hello @tsacks, any news?
